### PR TITLE
Submit AdMeta Milestone 1 Delivery

### PR DESCRIPTION
**Pull request by h4n0**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/731

### CODE REVIEW

**Comment by alxs** [comment]
Thanks for the delivery @h4n0. We'll look into it as soon as possible.

**Comment by 0xCaso** [comment]
Hi @h4n0! I'm starting the evaluation, I'll let you know if I need help!

File: deliveries/admeta_milestone_1.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**  
+
+* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/AdMeta.md
+* **Milestone Number:** 1
+
+
+**Deliverables**
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
```
Comment (0xCaso) [inline deliveries/admeta_milestone_1.md:12]: In the meantime, could you please add the license in the deliverables?
```suggestion
| ------------- | ------------- | ------------- |------------- |
| 0a. | License | [LICENSE](https://github.com/AdMetaNetwork/admeta/blob/main/LICENSE) | GPLv3 |
```

Comment (h4n0) [reply deliveries/admeta_milestone_1.md:12]: Oh sure! Thanks for the suggestion!

**Comment by 0xCaso** [comment]
Thanks for the edit.

I'm testing the webapp (and chain) locally; when I try to create an Ad Proposal, I get this error while I try to upload the image:
```
API resolved without sending a response for /api/upload, this may result in stalled requests.
写入成功
error - unhandledRejection: InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.
```
The node is running properly, and I transfered funds from Alice to the account I'm using.

Also, to link the front-end to the local node, I had to edit [this](https://github.com/AdMetaNetwork/admeta-webapp/blob/cb198db390708e47b1a3fce2b36769d01509e890/config/constant.ts#L6) file (I suggest you to write how to correctly setup the local environment in the README of admeta-webapp). I've only edited the `polkadot_network` const, but not the `HTTP_SERVER`.

Finally, I'm using the webapp repo at the commit you provided in the deliverables ([cb198db](https://github.com/AdMetaNetwork/admeta-webapp/commit/cb198db390708e47b1a3fce2b36769d01509e890)).

**Review by 0xCaso — COMMENTED** [review]
Thanks for the edit.

I'm testing the webapp (and chain) locally; when I try to create an Ad Proposal, I get this error while I try to upload the image:
```
API resolved without sending a response for /api/upload, this may result in stalled requests.
写入成功
error - unhandledRejection: InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.
```
The node is running properly, and I transfered funds from Alice to the account I'm using.

Also, to link the front-end to the local node, I had to edit [this](https://github.com/AdMetaNetwork/admeta-webapp/blob/cb198db390708e47b1a3fce2b36769d01509e890/config/constant.ts#L6) file (I suggest you to write how to correctly setup the local environment in the README of admeta-webapp). I've only edited the `polkadot_network` const, but not the `HTTP_SERVER`.

Finally, I'm using the webapp repo at the commit you provided in the deliverables ([cb198db](https://github.com/AdMetaNetwork/admeta-webapp/commit/cb198db390708e47b1a3fce2b36769d01509e890)).

**Comment by wxd56987** [comment]
> 
hi, @0xCaso , Our pictures are stored through Fleek, you need to register on [Fleek](https://fleek.co/) and generate a key, and replace the corresponding key in the config/c.example.ts file, don’t have to change the file name, /pages/api/upload.ts is the upload img api file. We will also describe clearly in the redeme that the network `polkadot_network = 'wss://testnet.admeta.network'`.

**Comment by 0xCaso** [comment]
Thanks. I'll let you know if this works, also I'd mention this thing in the README (having the Fleek account is required, where to set the key, ecc.)

**Comment by wxd56987** [comment]
Ok，I will add Fleek information to the README.

**Comment by wxd56987** [comment]
hi, @0xCaso Our project README already updated, You can watch this link [README](https://github.com/AdMetaNetwork/admeta-webapp/blob/testnet/README.md).

**Comment by 0xCaso** [comment]
Thanks a lot @wxd56987.

> Also, to link the front-end to the local node, I had to edit [this](https://github.com/AdMetaNetwork/admeta-webapp/blob/cb198db390708e47b1a3fce2b36769d01509e890/config/constant.ts#L6) file (I suggest you to write how to correctly setup the local environment in the README of admeta-webapp). I've only edited the polkadot_network const, but not the HTTP_SERVER.

Here I was meaning that you should add in the README where the dev can (and has to) put `ws://localhost:9944` instead of `wss://testnet.admeta.network` for the local development;

Also, in [`c.example.ts`](https://github.com/AdMetaNetwork/admeta-webapp/blob/testnet/config/c.example.ts), there are two vars for the Fleek's keys. Does `f_api_key` correspond to the Hosting API, and `f_api_secret` to the Storage API? Because I can't find the secret key inside my Fleek account (I can just generate Hosting and Storage API keys).

**Comment by wxd56987** [comment]
Hi @0xCaso ，When Fleek creates a key, it will also generate a secret, you can read this document https://docs.fleek.co/storage/fleek-storage-js/, please try again. I will modify the README right away.

**Comment by 0xCaso** [comment]
Many thanks @wxd56987, it works with the Fleek keys!

Now, how can I approve locally an `ad`? Should I do it with a chain `council`'s account? Because, for example, Alice should be part of it by default, but I can't approve ads with it.

**Review by 0xCaso — COMMENTED** [review]
Many thanks @wxd56987, it works with the Fleek keys!

Now, how can I approve locally an `ad`? Should I do it with a chain `council`'s account? Because, for example, Alice should be part of it by default, but I can't approve ads with it.

**Comment by wxd56987** [comment]
Hi @0xCaso ,Yes, the current permission of approve can only be operated by sudo. There is currently no operation entry on the web side, and the operation review must be performed in Polkadot.js . you can do it here https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Ftestnet.admeta.network#/extrinsics. Can you send me your discord account, I'll send you a screenshot

**Comment by 0xCaso** [comment]
Just managed to approve with `sudo`, thanks!

**Comment by 0xCaso** [comment]
Last question for the tutorial @wxd56987: is there an `ad url` I can put in the `ad` creation in order to test the "_Check Matched Ads and Claim Rewards_" section locally? If not, are there other ways to do it?

**Comment by h4n0** [comment]
Hi @0xCaso 

Use [this one](https://storageapi.fleek.one/cfc90be6-92dc-4a29-878e-bb564cc64b76-bucket/admeta_sample_billboard.jpeg) if you wanna test it with Polkadot.js App.

Basically all on IPFS stored images are supported here.

**Comment by 0xCaso** [comment]
I was meaning an `ad url` to put here:
Screenshot 2023-01-05 at 15 47 45
So not an url to use as the `ad image`, but an url that makes me test the "Claim Rewards" functionality.

Otherwise should I directly test it on Polkadot.js App?

**Comment by 0xCaso** [comment]
So should I just put that link (which should be an image) in the below field?
Screenshot 2023-01-05 at 15 47 45
And, if yes, what should I do in PolkadotJS App to complete the tasks and receive the rewards?

**Comment by 0xCaso** [comment]
Thanks a lot @h4n0 for the explaining, it makes much sense now!

By the way, locally I managed to view an `ad` that I previously added and approved. Then I wanted to test specifically the "_Claim Rewards_" functionality, is it possible or do I have to use your testnet or hosted webapp?

**Review by 0xCaso — COMMENTED** [review]
Thanks a lot @h4n0 for the explaining, it makes much sense now!

By the way, locally I managed to view an `ad` that I previously added and approved. Then I wanted to test specifically the "_Claim Rewards_" functionality, is it possible or do I have to use your testnet or hosted webapp?

**Comment by 0xCaso** [comment]
Ok thanks, I'll try with the hosted webapp then. Could you send some funds to `5G1QY5qEweAgUpZaWs23iC8B5ejKsFQQ5WAWuBwS5XLeSajt`?

**Comment by h4n0** [comment]
> 5G1QY5qEweAgUpZaWs23iC8B5ejKsFQQ5WAWuBwS5XLeSajt

Sure, it's funded now!

**Comment by 0xCaso** [comment]
Thanks! I managed to update my profile with the age in the range, the "DeFi" tag and "ad display" enabled but the ad is not visible (also after n>2 blocks)

**Review by 0xCaso — COMMENTED** [review]
Thanks! I managed to update my profile with the age in the range, the "DeFi" tag and "ad display" enabled but the ad is not visible (also after n>2 blocks)

**Comment by h4n0** [comment]
> Thanks! I managed to update my profile with the age in the range, the "DeFi" tag and "ad display" enabled but the ad is not visible (also after n>2 blocks)

Hi @0xCaso , I just checked the previously published ad and its total amount has been reduced to 0 due to my testing matches. I've created a new ad and now observed that your account has been matched with this one. You can now try to claim your reward and check your balance changes.

**Comment by h4n0** [comment]
> Thanks again. Now I see the ad, but if I click on "View ad and claim rewards", I get redirected to https://admeta.network/ and I cannot do anything else

Oh yes, that is expected behavior. Sorry I made a bad example with this link as a callback URL. Our design here is to have a properly designed callback page ([this page](https://click.admeta.network/) can be an example). The reason to use this page is to avoid users from claiming rewards without viewing or interacting with the ad content.

In the current MVP, we haven't implemented the feature "claiming with randomness parameter", so you can call the "claim rewards" extrinsic directly from the Polkadot.js App, under `user->claimReward` with params `proposer: 5FumNjmQzxBF5Zs6WYeS3QwVuUxbzWvoZyqENMNiGu1Vksym, adIndex: 2`

**Comment by 0xCaso** [comment]
Clear. Also https://click.admeta.network UI won't connect the PolkadotJS extension, but just Metamask, so it would have been impossible anyway.

About the "claiming with randomness parameter", that would be the functionality that enable users to actually do something to be able to claim rewards? And now everyone can claim rewards without doing any task, right?

**Comment by h4n0** [comment]
>Clear. Also https://click.admeta.network/ UI won't connect the PolkadotJS extension, but just Metamask, so it would have been impossible anyway.

That is right.

>About the "claiming with randomness parameter", that would be the functionality that enable users to actually do something to be able to claim rewards? And now everyone can claim rewards without doing any task, right?

Exactly, we plan to create a merkle tree for each published ad, and the merkle root stores on chain while the leaves of this merkle tree are random numbers stored offchain. So users cannot claim rewards by calling this extrinsic themselves.

**Review by 0xCaso — CHANGES_REQUESTED** [review]
Thanks for the explanations! I wrote the evaluation, you can read it [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/AdMeta_1_0xCaso.md). Let me know if you have any question!

**Comment by h4n0** [comment]
> Thanks for the explanations! I wrote the evaluation, you can read it [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/AdMeta_1_0xCaso.md). Let me know if you have any question!

Thanks a lot @0xCaso for the feedback! We'll work on the WebApp guide these days and come back to you shortly!

**Comment by h4n0** [comment]
Hi @0xCaso , we have updated our WebApp guide for testing in README by [this commit](https://github.com/AdMetaNetwork/admeta-webapp/blob/8804631c650a31223aba196702a8a6080883e4dd/README.md)

Please check and feel free to tell me if anything else needs to be improved.

**Review by 0xCaso — CHANGES_REQUESTED** [review]
Thanks a lot for the update. By the way, in the evaluation I was meaning you should add the explanation of how to test the dapp locally, not how to setup it (example: specify to accept the `ad` with `sudo` on PolkadotJS, specify to use `claimReward`, ecc...); View https://github.com/w3f/Grant-Milestone-Delivery/pull/667#issuecomment-1372120084 https://github.com/w3f/Grant-Milestone-Delivery/pull/667#issuecomment-1375711955

**Comment by h4n0** [comment]
> Thanks a lot for the update. By the way, in the evaluation I was meaning you should add the explanation of how to test the dapp locally, not how to setup it (example: specify to accept the `ad` with `sudo` on PolkadotJS, specify to use `claimReward`, ecc...); View [#667 (comment)](https://github.com/w3f/Grant-Milestone-Delivery/pull/667#issuecomment-1372120084) [#667 (comment)](https://github.com/w3f/Grant-Milestone-Delivery/pull/667#issuecomment-1375711955)

Yes, that makes sense. We updated the whole process for users to test our WebApp locally. Please find it in this commit 
https://github.com/AdMetaNetwork/admeta-webapp/tree/bb74d090d815bb9af5cfb9314968672bffa54ffa

**Review by 0xCaso — APPROVED** [review]
Thanks again for the changing, I updated the evaluation and for me the deliverable is fully acceptable. Let's wait the confirm from the team!

**Review by Noc2 — APPROVED** [review]

**Comment by 0xCaso** [comment]
Congrats for the grants completion!

One last thing @h4n0: recently we changed our guidelines, and I would ask you if it's possible to change USDT on Ethereum as payment method to another option choosable from [here](https://github.com/w3f/Grants-Program#pencil-process) (so if you want to stay with Ethereum you should change it to USDC or DAI).

If it's possible then, I'd ask you to resend the invoice, indicating here also the token you've chosen. Thanks!

**Comment by h4n0** [comment]
> Congrats for the grants completion!
> 
> One last thing @h4n0: recently we changed our guidelines, and I would ask you if it's possible to change USDT on Ethereum as payment method to another option choosable from [here](https://github.com/w3f/Grants-Program#pencil-process) (so if you want to stay with Ethereum you should change it to USDC or DAI).
> 
> If it's possible then, I'd ask you to resend the invoice, indicating here also the token you've chosen. Thanks!

Hi @0xCaso , thank you very much!!!

I've checked our invoice, and it seems we didn't specify USDT on the invoice, but just USD as only fiat currencies are supported during the invoice generation. The USDT was specified only in our Grant application. 

So we accept USDC as well, with the same reception ETH address. So should we reuse the old invoice or do you need a new one with USDC specified?

**Comment by 0xCaso** [comment]
> Hi @0xCaso , thank you very much!!!
> 
> I've checked our invoice, and it seems we didn't specify USDT on the invoice, but just USD as only fiat currencies are supported during the invoice generation. The USDT was specified only in our Grant application.
> 
> So we accept USDC as well, with the same reception ETH address. So should we reuse the old invoice or do you need a new one with USDC specified?

The old one is perfect, just replace in the "Notes" field "ETH" with "ETH-USDC"

**Comment by h4n0** [comment]
Thanks! Alright, the New invoice submitted!

**Comment by 0xCaso** [comment]
Thanks again, seems perfect now! I just notified the Finance team, please allow up to 14 days for the payment.

**Comment by RouvenP** [comment]
hi @h4n0 we transferred the payment today.

**Pull request merged**